src/main/java/com/nirima/jenkins/plugins/docker/DockerNodeFactory.java

```diff
@@ -0,0 +1,106 @@
+package com.nirima.jenkins.plugins.docker;
+
+import com.google.inject.internal.asm.$ClassVisitor;
+import com.nirima.jenkins.plugins.docker.cloudstat.CloudStatsFactory;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.model.Descriptor;
+import hudson.model.Node;
+import hudson.slaves.ComputerLauncher;
+import hudson.slaves.NodeProvisioner;
+import io.jenkins.docker.DockerTransientNode;
+import org.jenkinsci.plugins.cloudstats.ProvisioningActivity;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.concurrent.Future;
+
+public interface DockerNodeFactory extends Serializable  {
+
+    /**
+     * Safely load the instance of the DockerNodeFactory
+     * @return a new instance of a implementation of DockerNodeFactory
+     */
+    static DockerNodeFactory getInstance() {
+        try{
+            return (DockerNodeFactory)Class.forName("com.nirima.jenkins.plugins.docker.cloudstat.CloudStatsFactory").newInstance();
```

Comment:
can't we use `[@Extension(optional = true)](https://www.jenkins.io/doc/developer/plugin-development/optional-dependencies/)` annotation on CloudStatsFactory instead, and check for available descriptors?

Reply:
Yeah that looks like the correct approach. This was my first time working in a Jenkins plugin so please forgive the lack of Jenkins-ey-ness :) 

Regarding testing... I'm not sure that it is possible to unit test "without" a dependency marked optional, unless an external driver re-writes the POM from optional -> compile

Reply:
Also, I don't think our DevOps guys got my branch to work, and I didn't have time to debug... This might be why...